apps/mongodb.go

```diff
@@ -0,0 +1,291 @@
+	c := []fluentbit.Component{}
+
+	c = append(c, p.JsonLogComponents(tag, uid)...)
+	c = append(c, p.RegexLogComponents(tag, uid)...)
+
+	return c
+}
+
+// JsonLogComponents are the fluentbit components for parsing log messages that are json formatted.
+// these are generally messages from mongo with versions greater than or equal to 4.4
+func (p *LoggingProcessorMongodb) JsonLogComponents(tag, uid string) []fluentbit.Component {
+	c := p.jsonParserWithTimeKey(tag, uid)
+
+	c = append(c, p.promoteWiredTiger(tag, uid)...)
+	c = append(c, p.severityParser(tag, uid)...)
+	c = append(c, p.renames(tag, uid)...)
+
+	return c
+}
+
+// jsonParserWithTimeKey requires promotion of the nested timekey for the json parser so we must
+// first promote the $date field from the "t" field before declaring the parser
+func (p *LoggingProcessorMongodb) jsonParserWithTimeKey(tag, uid string) []fluentbit.Component {
+	c := []fluentbit.Component{}
+
+	jsonParser := &confgenerator.LoggingProcessorJsonParser{
+		ParserShared: confgenerator.ParserShared{
+			TimeKey:    "time",
+			TimeFormat: "%Y-%m-%dT%H:%M:%S.%L%z",
+			Types: map[string]string{
+				"id":      "integer",
+				"message": "string",
+			},
+		},
+	}
+	jpComponents := jsonParser.Components(tag, uid)
+
+	parserComponent, filterComponent := jpComponents[0], jpComponents[1]
+
+	c = append(c, parserComponent, filterComponent)
+
+	tempPrefix := "temp_ts_"
+	timeKey := "time"
+	// have to bring $date to top level in order for it to be parsed as timeKey
+	// see https://github.com/fluent/fluent-bit/issues/1013
+	liftTs := fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":         "nest",
+			"Match":        tag,
+			"Operation":    "lift",
+			"Nested_under": "t",
+			"Add_prefix":   tempPrefix,
+		},
+	}
+
+	renameTsOption := fluentbit.NewHardRenameModifyOptions(fmt.Sprintf("%s$date", tempPrefix), timeKey, nil)
+	renameTs := renameTsOption.Component(tag)
+
+	c = append(c, liftTs, renameTs)
+
+	// IMPORTANT: now that we have lifted the json to top level
+	// we need to re-parse in order to properly set time at the
+	// parser level
+	c = append(c, filterComponent)
+
+	removeTimestamp := fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Name":   "modify",
+			"Match":  tag,
+			"Remove": timeKey,
+		},
+	}
+	c = append(c, removeTimestamp)
+
+	return c
+}
+
+func (p *LoggingProcessorMongodb) severityParser(tag, uid string) []fluentbit.Component {
+	severityComponents := []fluentbit.Component{}
+	severityKey := "logging.googleapis.com/severity"
+
+	severityComponents = append(severityComponents, fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Match":  tag,
+			"Name":   "modify",
+			"Rename": "s severity",
+		},
+	})
+
+	severityComponents = append(severityComponents, fluentbit.TranslationComponents(tag, "severity", severityKey, []struct {
+		SrcVal  string
+		DestVal string
+	}{
+		{
```

Comment:
[Optional] We could compress this section to save some space and improve readability. Similar to https://github.com/GoogleCloudPlatform/ops-agent/blob/d88a768910a7f4eb2a4bbae94bb829ea9effa261/apps/nginx.go#L146-L158

Reply:
Ahh I misunderstood when i first read, you just want to omit the extra whitespace/unneeeded declarations. Sorry about that!